code/modules/hydroponics/gene_modder.dm

```diff
@@ -379,38 +255,103 @@
+				update_icon(UPDATE_OVERLAYS)
+			else
+				var/obj/item/I = user.get_active_hand()
+				if(istype(I, /obj/item/seeds))
+					add_seed(I, user)
+			SStgui.update_uis(src)
+		if("eject_disk")
+			if(disk)
+				disk.forceMove(loc)
+				user.put_in_hands(disk)
+				disk = null
+				update_genes()
+			else
+				var/obj/item/I = usr.get_active_hand()
+				if(istype(I, /obj/item/disk/plantgene))
+					add_disk(I, user)
+			SStgui.update_uis(src)
+		if("variant_name")
+			seed.variant_prompt(usr, src)
+			SStgui.update_uis(src)
+			// Todo, maybe someday, make this into a modal and make it look good
+
+		if("extract")
+			var/dat = "Are you sure you want to extract [target.get_name()] gene from the [seed]? The sample will be destroyed in process! "
+			if(istype(target, /datum/plant_gene/core))
+				var/datum/plant_gene/core/gene = target
+				if(istype(target, /datum/plant_gene/core/potency))
+					if(gene.value > max_potency)
+						dat += "This device's extraction capabilities are currently limited to [max_potency] potency. "
+						dat += "Target gene will be degraded to [max_potency] potency on extraction."
+				else if(istype(target, /datum/plant_gene/core/lifespan))
+					if(gene.value > max_endurance)
+						dat += "This device's extraction capabilities are currently limited to [max_endurance] lifespan. "
+						dat += "Target gene will be degraded to [max_endurance] Lifespan on extraction."
+				else if(istype(target, /datum/plant_gene/core/endurance))
+					if(gene.value > max_endurance)
+						dat += "This device's extraction capabilities are currently limited to [max_endurance] endurance. "
+						dat += "Target gene will be degraded to [max_endurance] endurance on extraction."
+				else if(istype(target, /datum/plant_gene/core/yield))
+					if(gene.value > max_yield)
+						dat += "This device's extraction capabilities are currently limited to [max_yield] yield. "
+						dat += "Target gene will be degraded to [max_yield] yield on extraction."
+				else if(istype(target, /datum/plant_gene/core/production))
+					if(gene.value < min_production)
+						dat += "This device's extraction capabilities are currently limited to [min_production] production. "
+						dat += "Target gene will be degraded to [min_production] production on extraction."
+				else if(istype(target, /datum/plant_gene/core/weed_rate))
+					if(gene.value < min_wrate)
+						dat += "This device's extraction capabilities are currently limited to [min_wrate] weed rate. "
+						dat += "Target gene will be degraded to [min_wrate] weed rate on extraction."
+				else if(istype(target, /datum/plant_gene/core/weed_chance))
+					if(gene.value < min_wchance)
+						dat += "This device's extraction capabilities are currently limited to [min_wchance] weed chance. "
+						dat += "Target gene will be degraded to [min_wchance] weed chance on extraction."
+			ui_modal_boolean(src, action, dat, yes_text = "Extract", no_text = "Cancel", delegate_no = PROC_REF(do_nothing))
```

Comment:
```suggestion
			ui_modal_boolean(src, action, dat, yes_text = "Extract", no_text = "Cancel")
```
you don't need this I'm pretty sure in any of these `ui_modal_boolean()` calls
```dm
/datum/ui_modal/boolean/on_answer(answer)
	if(answer)
		return ..(answer, arguments)
	else if(delegate_no)
		return call(owning_source, delegate_no)(arguments)
	return FALSE
```
based on your code and `do_nothing` the same things are done regardless if you put in a value for `delegate_no` or not

Reply:
It needs to be there, otherwise it carries through with the answering function and doesn't delegate. If you press the cancel button it will continue through with the operation

Reply:
ok, instead of specifying `delegate_no` just specify `delegate` and make it call setter process
i.e.
```dm
/proc/gene_replace()
	if(disk && disk.gene && istype(disk.gene, target.type) && istype(target, /datum/plant_gene/core))
		seed.genes -= target
		var/datum/plant_gene/core/C = disk.gene.Copy()
		seed.genes += C
		C.apply_stat(seed)
		repaint_seed()

//and.....
/ui_act()
	if("replace")
		ui_modal_boolean(src, action, "Are you sure you want to replace [target.get_name()] gene with [disk.gene.get_name()]?", yes_text = "Replace", no_text = "Cancel", delegate = PROC_REF(gene_replace))
```
this is the proper way to go about what you're trying to do. Modals work in cycles where they make another run around if you perform an input action on the modal unless you specify an escape option (such as delegating to a proc call). In this case, we're done with the modal TGUI if we hit either button so we immediately want to jump into DM stuff instead of nesting the logic further into Modals.

This will actually help cleanup your code and make it more reusable by pulling this action functionality out of the UI into just a method on the gene modder type.